proving/proving.go

```
@@ -0,0 +1,303 @@
+	if err != nil {
+		return nil, err
+	}
+	gsReader := shared.NewGranSpecificReader(reader, p.cfg.BitsPerLabel)
+
+	numWorkers := endNonce - startNonce + 1
+	var indices []byte
+
+	workersChans := make([]chan []byte, numWorkers)
+	for i := range workersChans {
+		workersChans[i] = make(chan []byte, 1000) // TODO(moshababo): use numLabels/2 size instead? need just enough buffer to circulate between the two routines
+	}
+	resultsChan := make(chan nonceResult, numWorkers)
+	errChan := make(chan error, numWorkers)
+
+	// Start IO worker.
+	// Feed all labels into each worker chan.
+	go func() {
+		for {
+			label, err := gsReader.ReadNext()
+			if err != nil {
+				for i := range workersChans {
+					close(workersChans[i])
+				}
+
+				if err != io.EOF {
+					errChan <- err
+				}
+				break
+			}
+
+			for i := range workersChans {
+				workersChans[i] <- label
+			}
+		}
+	}()
+
+	// Start a worker for each nonce.
+	ctx, cancel := context.WithCancel(context.Background())
+	for i := uint32(0); i < numWorkers; i++ {
+		i := i
+		go func() {
+			nonce := startNonce + i
+			indices, err = p.tryNonce(ctx, numLabels, challenge, nonce, workersChans[i], difficulty)
+			resultsChan <- nonceResult{nonce, indices, err}
+		}()
+	}
+
+	// Drain the workers results chan.
+	var solutionNonceResult *nonceResult
+	for i := uint32(0); i < numWorkers; i++ {
+		res := <-resultsChan
+		if res.err != nil {
+			p.logger.Debug("proving: nonce %v failed: %v", res.nonce, res.err)
+		} else {
+			p.logger.Debug("proving: nonce %v succeeded", res.nonce)
+			cancel()
+
+			// There might be multiple successful nonces due to race condition with the cancellation,
+			// but this is not a problem. We'll use the last one to arrive.
+			solutionNonceResult = &res
+		}
+	}
+
+	// Check for an error from the IO worker.
+	select {
+	case err := <-errChan:
+		p.logger.Debug("proving: error: %v", err)
+		return nil, err
+	default:
+	}
```

If an io error occurred, we don't necessarily want to return an error. Because the worker chans are buffered, one of the workers may have succeeded after the error occurred, and then we want to return the successful result.

---

Will be addressed in a follow-up PR.